### ISSUE TITLE
A Simple Explanation of the Softmax Function

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4870]
A Simple Explanation of the Softmax Function<br>
<br>
https://ift.tt/2Oi4PY3<br>
<br>
<br>
<br>
<div><div>
<p><strong>Softmax turns arbitrary real values into probabilities</strong>, which are often useful in Machine Learning. The math behind it is pretty simple: given some numbers,</p>
<ol>
<li>Raise <a rel="noopener noreferrer" href="https://en.wikipedia.org/wiki/E_(mathematical_constant)" target="_blank">e</a> (the mathematical constant) to the power of each of those numbers.</li>
<li>Sum up all the exponentials (powers of <span><span></span><span><span><span></span><span>e</span></span></span></span>). This result is the <em>denominator</em>.</li>
<li>Use each number’s exponential as its <em>numerator</em>.</li>
<li>
<span><span></span><span><span><span></span><span><span>Probability</span></span><span></span><span>=</span><span></span></span><span><span></span><span><span></span><span><span><span><span><span><span></span><span><span><span><span>Denominator</span></span></span></span></span><span><span></span><span></span></span><span><span></span><span><span><span><span>Numerator</span></span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span></span></span></span></span></span>.</li>
</ol>
<p>Written more fancily, Softmax performs the following transform on <span><span></span><span><span><span></span><span>n</span></span></span></span> numbers <span><span></span><span><span><span></span><span><span>x</span><span><span><span><span><span><span></span><span><span>1</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span></span><span>…</span><span></span><span><span>x</span><span><span><span><span><span><span></span><span><span>n</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span></span>:</p>
<span><span><span></span><span><span><span></span><span>s</span><span>(</span><span><span>x</span><span><span><span><span><span><span></span><span><span>i</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span>)</span><span></span><span>=</span><span></span></span><span><span></span><span><span></span><span><span><span><span><span><span></span><span><span><span>∑</span><span><span><span><span><span><span></span><span><span><span>j</span><span>=</span><span>1</span></span></span></span><span><span></span><span><span>n</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span><span></span><span><span>e</span><span><span><span><span><span><span></span><span><span><span><span>x</span><span><span><span><span><span><span></span><span><span>j</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span></span></span></span></span></span></span></span></span><span><span></span><span></span></span><span><span></span><span><span><span>e</span><span><span><span><span><span><span></span><span><span><span><span>x</span><span><span><span><span><span><span></span><span><span>i</span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span></span></span></span></span></span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span></span></span></span></span></span></span>
<p>The outputs of the Softmax transform are always in the range <span><span></span><span><span><span></span><span>[</span><span>0</span><span>,</span><span></span><span>1</span><span>]</span></span></span></span> and add up to 1. Hence, they form a <strong>probability distribution</strong>.</p>
<h2>
<a href="#a-simple-example"></a>A Simple Example</h2>
<p>Say we have the numbers -1, 0, 3, and 5. First, we calculate the denominator:</p>
<span><span><span></span><span><span><span></span><span><span><span><span><span><span><span><span></span><span><span><span>Denominator</span></span></span></span><span><span></span><span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span><span><span><span><span><span></span><span><span></span><span></span><span>=</span><span></span><span><span>e</span><span><span><span><span><span><span></span><span><span><span>−</span><span>1</span></span></span></span></span></span></span></span></span><span></span><span>+</span><span></span><span><span>e</span><span><span><span><span><span><span></span><span><span>0</span></span></span></span></span></span></span></span><span></span><span>+</span><span></span><span><span>e</span><span><span><span><span><span><span></span><span><span>3</span></span></span></span></span></span></span></span><span></span><span>+</span><span></span><span><span>e</span><span><span><span><span><span><span></span><span><span>5</span></span></span></span></span></span></span></span></span></span><span><span></span><span><span></span><span></span><span>=</span><span></span><span><span><span><span><span><span></span><span><span><span>1</span><span>6</span><span>9</span><span>.</span><span>8</span><span>7</span></span></span></span><span><span></span><span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span></span></span></span></span></span></span>
<p>Then, we can calculate the numerators and probabilities:</p>
<table>
<thead>
<tr>
<th><span><span></span><span><span><span></span><span>x</span></span></span></span></th>
<th>Numerator (<span><span></span><span><span><span></span><span><span>e</span><span><span><span><span><span><span></span><span><span>x</span></span></span></span></span></span></span></span></span></span></span>)</th>
<th>Probability (<span><span></span><span><span><span></span><span><span></span><span><span><span><span><span><span></span><span><span><span>1</span><span>6</span><span>9</span><span>.</span><span>8</span><span>7</span></span></span></span><span><span></span><span></span></span><span><span></span><span><span><span><span>e</span><span><span><span><span><span><span></span><span><span>x</span></span></span></span></span></span></span></span></span></span></span></span><span>​</span></span><span><span><span></span></span></span></span></span><span></span></span></span></span></span>)</th>
</tr>
</thead>
<tbody>
<tr>
<td>-1</td>
<td>0.368</td>
<td>0.002</td>
</tr>
<tr>
<td>0</td>
<td>1</td>
<td>0.006</td>
</tr>
<tr>
<td>3</td>
<td>20.09</td>
<td>0.118</td>
</tr>
<tr>
<td>5</td>
<td>148.41</td>
<td>0.874</td>
</tr>
</tbody>
</table>
<p>The bigger the <span><span></span><span><span><span></span><span>x</span></span></span></span>, the higher its probability. Also, notice that the probabilities all add up to 1, as mentioned before.</p>
<h2>
<a href="#implementing-softmax-in-python"></a>Implementing Softmax in Python</h2>
<p>Using <a rel="noopener noreferrer" href="https://www.numpy.org/" target="_blank">numpy</a> makes this super easy:</p>
<div data-language="python"><pre><code><span>import</span> numpy <span>as</span> np

<span>def</span> <span>softmax</span><span>(</span>xs<span>)</span><span>:</span>
    <span>return</span> np<span>.</span>exp<span>(</span>xs<span>)</span> <span>/</span> <span>sum</span><span>(</span>np<span>.</span>exp<span>(</span>xs<span>)</span><span>)</span>

xs <span>=</span> np<span>.</span>array<span>(</span><span>[</span><span>-</span><span>1</span><span>,</span> <span>0</span><span>,</span> <span>3</span><span>,</span> <span>5</span><span>]</span><span>)</span>
<span>print</span><span>(</span>softmax<span>(</span>xs<span>)</span><span>)</span> </code></pre></div>
<figcaption><a href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.exp.html" target="_blank">np.exp()</a> raises e to the power of each element in the input array.</figcaption>
<h2>
<a href="#why-is-softmax-useful"></a>Why is Softmax useful?</h2>
<p>Imagine building a <a href="https://victorzhou.com/blog/intro-to-neural-networks/">Neural Network</a> to answer the question: <span>Is this picture of a dog or a cat?</span></p>
<p>A common design for this neural network would have it output 2 real numbers, one representing <em>dog</em> and the other <em>cat</em>, and apply Softmax on these values. For example, let’s say the network outputs <span><span></span><span><span><span></span><span>[</span><span>−</span><span>1</span><span>,</span><span></span><span>2</span><span>]</span></span></span></span>:</p>
<table>
<thead>
<tr>
<th>Animal</th>
<th><span><span></span><span><span><span></span><span>x</span></span></span></span></th>
<th><span><span></span><span><span><span></span><span><span>e</span><span><span><span><span><span><span></span><span><span>x</span></span></span></span></span></span></span></span></span></span></span></th>
<th>Probability</th>
</tr>
</thead>
<tbody>
<tr>
<td>Dog</td>
<td>-1</td>
<td>0.368</td>
<td>0.047</td>
</tr>
<tr>
<td>Cat</td>
<td>2</td>
<td>7.39</td>
<td>0.953</td>
</tr>
</tbody>
</table>
<p>This means our network is <strong>95.3% confident</strong> that the picture is of a cat. Softmax lets us <strong>answer classification questions with probabilities</strong>, which are more useful than simpler answers (e.g. binary yes/no).</p>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via victorzhou.com https://victorzhou.com<br>
<br>
July 26, 2019 at 11:07AM
